src/citrine/ara/variables.py

```diff
@@ -0,0 +1,98 @@
+"""Variable definitions for Ara."""
+from typing import Type, Optional, List  # noqa: F401
+from abc import abstractmethod
+
+from taurus.entity.link_by_uid import LinkByUID
+
+from citrine._serialization.serializable import Serializable
+from citrine._serialization.polymorphic_serializable import PolymorphicSerializable
+from citrine._serialization import properties
+
+
+class Variable(PolymorphicSerializable['Variable']):
+    """A variable that can be assigned values present in material histories.
+
+    Abstract type that returns the proper type given a serialized dict.
+    """
+
+    @abstractmethod
+    def _attrs(self) -> List[str]:
+        pass  # pragma: no cover
+
+    def __eq__(self, other):
+        try:
+            return all([
+                self.__getattribute__(key) == other.__getattribute__(key) for key in self._attrs()
+            ])
+        except AttributeError:
+            return False
+
+    @classmethod
+    def get_type(cls, data) -> Type[Serializable]:
+        """Return the subtype."""
+        types: List[Type[Serializable]] = [
+            RootInfo, AttributeByTemplate
+        ]
+        return next(x for x in types if x.type == data["type"])
```

Comment:
I think this could be more defensive, throwing exceptions if `data` is missing a "type" field or if it doesn't match the `types` list (say, because something new got added in the backend but not in citrine-python).

I looked at the other implementations of `get_type`, and they're all doing basically the same thing, but there are several different styles and levels of error checking. Unrelated to this PR, I think we could require each `PolymorphicSerializable` to have some "types dictionary" and make the `get_type` method more generic. This would also allow for more error checking without having to write tests for that error checking for every single implementation of PolymorphicSerializable.

I can make the issue/ticket and take care of it if you think it's worthwhile.

Reply:
I can update it here, but I don't want to broaden the scope of this branch to fixing the pattern everywhere.  :+1: on creating an issue for it.  It probably relates to https://citrine.atlassian.net/browse/PLA-2444

Reply:
2444 is also related to deserialization, but I think that 2444 is significantly different in spirit (it focuses on complications around traversing a tree of data concepts objects).

Agree about not broadening the scope, and I don't even think you need to update the code in this PR, since what you wrote adheres to established patterns. I'll make an issue.